pkg/cleanup/controller/controller.go

```diff
@@ -0,0 +1,302 @@
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// PVC was deleted in the meantime, ignore.
+			klog.Infof("PVC %q in namespace %q no longer exists", pvClaimRef.Name, pvClaimRef.Namespace)
+			return nil
+		}
+		return err
+	}
+	// Check that the PVC we're about to delete still points back to the PV that enqueued it.
+	if pvc.Spec.VolumeName != pv.Name {
+		klog.Infof("PVC %q no longer references PV %q so will not be cleaned up", pvc.Name, pv.Name)
+		return nil
+	}
+
+	err = c.deletePVC(ctx, pvc)
+	if err != nil {
+		klog.Errorf("failed to delete pvc %q in namespace &q: %w", pvClaimRef.Name, pvClaimRef.Namespace, err)
+		return err
+	}
+
+	klog.Infof("Deleted PVC %q that pointed to Node %q", pvClaimRef.Name, nodeName)
+	return nil
+}
+
+func (c *CleanupController) nodeDeleted(obj interface{}) {
+	c.startCleanupTimersIfNeeded()
+}
+
+// startCleanupTimersIfNeeded enqueues any local PVs
+// with a NodeAffinity to a deleted Node and a StorageClass listed in storageClassNames.
+func (c *CleanupController) startCleanupTimersIfNeeded() {
+	pvs, err := c.pvLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("error listing pvs: %w", err)
+		return
+	}
+
+	for _, pv := range pvs {
+		nodeName, ok := common.NodeAttachedToLocalPV(pv)
+		if !ok {
+			klog.Errorf("error getting node attached to pv: %s", pv)
+			continue
+		}
+
+		shouldEnqueue, err := c.shouldEnqueueEntry(pv, nodeName)
+		if err != nil {
+			klog.Errorf("error determining whether to enqueue entry with pv %q: %w", pv.Name, err)
+			continue
+		}
+
+		if shouldEnqueue {
+			klog.Infof("Starting timer for resource deletion, resource:%s, timer duration: %s", pv.Spec.ClaimRef, c.pvcDeletionDelay.String())
+			c.eventRecorder.Event(pv.Spec.ClaimRef, v1.EventTypeWarning, "ReferencedNodeDeleted", fmt.Sprintf("PVC is tied to a deleted Node. PVC will be cleaned up in %s if the Node doesn't come back", c.pvcDeletionDelay.String()))
+
+			c.pvQueue.AddAfter(pv.Name, c.pvcDeletionDelay)
+		}
+	}
+}
+
+// shouldEnqueuePV checks if a PV should be enqueued to the entryQueue.
+// The PV must be a local PV, have a StorageClass present in the list of storageClassNames, have a NodeAffinity
+// to a deleted Node, and have a PVC bound to it (otherwise there's nothing to clean up).
+func (c *CleanupController) shouldEnqueueEntry(pv *v1.PersistentVolume, nodeName string) (bool, error) {
+	if !common.IsLocalPVWithStorageClass(pv, c.storageClassNames) || pv.Spec.ClaimRef == nil {
+		return false, nil
+	}
+
+	exists, err := common.NodeExists(c.nodeLister, nodeName)
+	return !exists && err == nil, err
+}
+
+// deletePVC deletes the PVC with the given name and namespace
+// and returns nil if the operation was successful or if the PVC doesn't exist
+func (c *CleanupController) deletePVC(ctx context.Context, pvc *v1.PersistentVolumeClaim) error {
+	err := c.client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(ctx, pvc.Name, metav1.DeleteOptions{})
```

Comment:
To prevent a possible conflict where someone else deletes and recreates the PVC at the same time we try to, it would be good to add the UID precondition when deleting the object: https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/blob/95d9b6325bb51f83efd7f5516a28086dcefddcf6/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L719

Reply:
Done